packages/panzer/disc-fe/src/Panzer_SubcellConnectivity.cpp

```diff
@@ -42,20 +42,82 @@
 
 #include "Panzer_SubcellConnectivity.hpp"
 
+#include "Teuchos_Assert.hpp"
+
 #include "Panzer_LocalMeshInfo.hpp"
 
 namespace panzer
 {
 
+int
+SubcellConnectivity::
+numSubcellsOnCell(const int cell) const
+{
+#ifndef NDEBUG
+  TEUCHOS_ASSERT(cell >= 0 and cell < _num_cells);
+#endif
```

Comment:
Instead of using NDEBUG, each Trilinos package has a define <package_name>_DEBUG so that we can selectively enable debug checks on a per-package basis. Debug for all trilinos packages tends to make debug builds way too slow. We should switch to PANZER_DEBUG instead of NDEBUG.

If you prefer to still use NDEBUG, a regular assert requires NDEBUG, so I would just use that instead of protecting the TEUCHOS_ASSERT with the ifdef NDEBUG. All of these instances in this file could be trimmed to a single line instead of 3 lines in this case. 

I noticed that some throws only occur in debug mode when NDEBUG is defined, but did not see any TEST_THROW() checks wrapped in #ifdef NDEBUG. Was this change tested in both debug and release builds?

Reply:
Awesome, I wasn't sure what the proper terms was. I usually wrap these in a range check define, but I wasn't sure if panzer had one of those. I'll switch these over to PANZER_DEBUG.